docs/specification/new-spec/blockchain.md

```diff
@@ -0,0 +1,389 @@
+and other results from the application.
+Elements of an object are accessed as expected,
+ie. `block.Header`. See [here](state.md) for the definition of `state`.
+
+## Header
+
+A Header is valid if its corresponding fields are valid.
+
+### Version
+
+Arbitrary string.
+
+### ChainID
+
+Arbitrary constant string.
+
+### Height
+
+```
+block.Header.Height > 0
+block.Header.Height == prevBlock.Header.Height + 1
+```
+
+The height is an incrementing integer. The first block has `block.Header.Height == 1`.
+
+### Time
+
+The median of the timestamps of the valid votes in the block.LastCommit.
+Corresponds to the number of nanoseconds, with millisecond resolution, since January 1, 1970.
+Increments with every new block.
+
+### NumTxs
+
+```
+block.Header.NumTxs == len(block.Txs)
+```
+
+Number of transactions included in the block.
+
+### TxHash
+
+```
+block.Header.TxHash == SimpleMerkleRoot(block.Txs)
+```
+
+Simple Merkle root of the transactions in the block.
+
+### LastCommitHash
+
+```
+block.Header.LastCommitHash == SimpleMerkleRoot(block.LastCommit)
+```
+
+Simple Merkle root of the votes included in the block.
+These are the votes that committed the previous block.
+
+The first block has `block.Header.LastCommitHash == []byte{}`
+
+### TotalTxs
+
+```
+block.Header.TotalTxs == prevBlock.Header.TotalTxs + block.Header.NumTxs
+```
+
+The cumulative sum of all transactions included in this blockchain.
+
+The first block has `block.Header.TotalTxs = block.Header.NumberTxs`.
+
+### LastBlockID
+
+```
+parts := MakeParts(block, state.ConsensusParams.BlockGossip.BlockPartSize)
+block.HeaderLastBlockID == BlockID{
+    SimpleMerkleRoot(block.Header),
+    PartsHeader{
+        SimpleMerkleRoot(parts),
+        len(parts),
+    },
+}
+```
+
+Previous block's BlockID. Note it depends on the ConsensusParams,
+which are held in the `state` and may be updated by the application.
+
+The first block has `block.Header.LastBlockID == BlockID{}`.
+
+### ResultsHash
+
+```
+block.ResultsHash == SimpleMerkleRoot(state.LastResults)
+```
+
+Simple Merkle root of the results of the transactions in the previous block.
+
+The first block has `block.Header.ResultsHash == []byte{}`.
+
+### AppHash
+
+```
+block.AppHash == state.AppHash
+```
+
+Arbitrary byte array returned by the application after executing and commiting the previous block.
+
+The first block has `block.Header.AppHash == []byte{}`.
+
+### ValidatorsHash
+
+```
+block.ValidatorsHash == SimpleMerkleRoot(state.Validators)
+```
+
+Simple Merkle root of the current validator set that is committing the block.
+This can be used to validate the `LastCommit` included in the next block.
+May be updated by the application.
+
+### ConsensusParamsHash
+
+```
+block.ValidatorsHash == SimpleMerkleRoot(state.ConsensusParams)
+```
+
+Simple Merkle root of the consensus parameters.
+May be updated by the application.
+
+### Proposer
+
+```
+block.Header.Proposer in state.Validators
```

Comment:
There are no Proposer field in block.Header. What do you mean with this?

Reply:
Is it good idea adding Proposer to the blockchain header? In general, I am not sure if adding more consensus stuff into blockchain header makes sense. It makes it more complex to validate, so it's harder job for light client for example. In principle, you only care about validator set and that enough of them sign that they commit to some block.

Reply:
I tend to agree, but it's a frequent feature request. That said, it's not something we intend for light clients to be able to authenticate easily any time soon.

It's also the "original proposer for the block", not even the proposer for the round the block got committed. So I'm not sure how useful it is. See comments here https://github.com/tendermint/tendermint/issues/952 and perhaps weight in :)

Reply:
I keep overlooking the reward aspect of blockchain. Although as you said, I am not sure original proposer of the block really matters.